presto-orc/src/main/java/com/facebook/presto/orc/writer/SegmentedSliceBlockBuilder.java

```diff
@@ -0,0 +1,416 @@
+    }
+
+    @Override
+    public BlockBuilder writeBytes(Slice source, int sourceIndex, int length)
+    {
+        if (openSegmentOffset == 0) {
+            // Expand Segments if necessary.
+            if (openSegment >= offsets.length) {
+                int newCapacity = Math.max(openSegment + 1, (int) (offsets.length * 1.5));
+                closedSlices = Arrays.copyOf(closedSlices, newCapacity);
+                offsets = Arrays.copyOf(offsets, newCapacity);
+            }
+
+            if (offsets[openSegment] == null) {
+                offsets[openSegment] = new int[SegmentHelper.SEGMENT_SIZE + 1];
+            }
+        }
+        openSliceOutput.writeBytes(source, sourceIndex, length);
+        return this;
+    }
+
+    @Override
+    public BlockBuilder closeEntry()
+    {
+        openSegmentOffset++;
+        offsets[openSegment][openSegmentOffset] = openSliceOutput.size();
+        if (openSegmentOffset == SegmentHelper.SEGMENT_SIZE) {
+            // Add the current finalized slice to closedSlices
+            Slice slice = openSliceOutput.copySlice();
+            closedSlices[openSegment] = slice;
+            closedSlicesSizeInBytes += slice.length();
+            closedSlicesRetainedSize += slice.getRetainedSize();
+
+            // Prepare the open slice for next write.
+            openSliceOutput.reset();
+            openSegmentOffset = 0;
+            openSegment++;
+        }
+        return this;
+    }
+
+    @Override
+    public BlockBuilder appendNull()
+    {
+        throw new UnsupportedOperationException(getClass().getName());
+    }
+
+    @Override
+    public BlockBuilder readPositionFrom(SliceInput input)
+    {
+        throw new UnsupportedOperationException(getClass().getName());
+    }
+
+    @Override
+    public boolean mayHaveNull()
+    {
+        return false;
+    }
+
+    @Override
+    protected boolean isEntryNull(int position)
+    {
+        return false;
+    }
+
+    @Override
+    public Block getRegion(int positionOffset, int length)
+    {
+        throw new UnsupportedOperationException(getClass().getName());
+    }
+
+    @Override
+    public Block copyRegion(int positionOffset, int length)
+    {
+        throw new UnsupportedOperationException(getClass().getName());
+    }
+
+    @Override
+    public Block build()
+    {
+        // No implementation of Segmented Slice based block exists and no usecase exists.
+        throw new UnsupportedOperationException("Not yet implemented");
+    }
+
+    @Override
+    public BlockBuilder newBlockBuilderLike(BlockBuilderStatus blockBuilderStatus)
+    {
+        return newBlockBuilderLike(blockBuilderStatus, getPositionCount());
+    }
+
+    @Override
+    public BlockBuilder newBlockBuilderLike(BlockBuilderStatus blockBuilderStatus, int expectedEntries)
+    {
+        if (blockBuilderStatus != null) {
+            throw new UnsupportedOperationException("Not yet implemented");
+        }
```

Comment:
`blockBuilderStatus` is actually fairly important. QQ: Is `newBlockBuilderLike` used anywhere in orc package? If not, just throw?

Reply:
This blockBuilder is only used by the SliceDictionaryBuilder and it passes in null for the blockBuilderStatus.

https://github.com/prestodb/presto/blob/master/presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryBuilder.java#L51